File: .github/workflows/update-supported-enterprise-server-versions/update.py
#!/usr/bin/env python3
import datetime
import json
import os
import pathlib

import semver

_API_COMPATIBILITY_PATH = pathlib.Path(__file__).absolute().parents[3] / "src" / "api-compatibility.json"
_ENTERPRISE_RELEASES_PATH = pathlib.Path(os.environ["ENTERPRISE_RELEASES_PATH"])
_RELEASE_FILE_PATH = _ENTERPRISE_RELEASES_PATH / "releases.json"
_FIRST_SUPPORTED_RELEASE = semver.VersionInfo.parse("2.22.0") # Versions older than this did not include Code Scanning.

def main():
	releases = json.loads(_RELEASE_FILE_PATH.read_text())
	oldest_supported_release = None
	newest_supported_release = None

	for release_version_string, release_data in releases.items():
		release_version = semver.VersionInfo.parse(release_version_string + ".0")
		if release_version < _FIRST_SUPPORTED_RELEASE:
			continue

		if newest_supported_release is None or release_version > newest_supported_release:
			feature_freeze_date = datetime.date.fromisoformat(release_data["feature_freeze"])
			if feature_freeze_date < datetime.date.today() + datetime.timedelta(weeks=2):
				newest_supported_release = release_version

		if oldest_supported_release is None or release_version < oldest_supported_release:
			end_of_life_date = datetime.date.fromisoformat(release_data["end"])
			if end_of_life_date > datetime.date.today():
				oldest_supported_release = release_version

	api_compatibility_data = {
		"minimumVersion": f"{oldest_supported_release.major}.{oldest_supported_release.minor}",
		"maximumVersion": f"{newest_supported_release.major}.{newest_supported_release.minor}",
	}
	_API_COMPATIBILITY_PATH.write_text(json.dumps(api_compatibility_data, sort_keys=True) + "\n")

if __name__ == "__main__":
	main()

File: .github/workflows/update-supported-enterprise-server-versions/update.ts
#!/usr/bin/env ts-node

import * as fs from 'fs';
import * as path from 'path';
import * as semver from 'semver';

const API_COMPATIBILITY_PATH = path.join(__dirname, '..', '..', '..', 'src', 'api-compatibility.json');
const ENTERPRISE_RELEASES_PATH = process.env.ENTERPRISE_RELEASES_PATH!;
const RELEASE_FILE_PATH = path.join(ENTERPRISE_RELEASES_PATH, 'releases.json');
const FIRST_SUPPORTED_RELEASE = semver.parse('2.22.0')!; // Versions older than this did not include Code Scanning.

interface ReleaseData {
  feature_freeze: string;
  end: string;
}

interface ApiCompatibility {
  minimumVersion: string;
  maximumVersion: string;
}

function main(): void {
  const apiCompatibilityData: ApiCompatibility = JSON.parse(fs.readFileSync(API_COMPATIBILITY_PATH, 'utf8'));
  const releases: Record<string, ReleaseData> = JSON.parse(fs.readFileSync(RELEASE_FILE_PATH, 'utf8'));

  // Remove GHES version using a previous version numbering scheme.
  if ('11.10' in releases) {
    delete releases['11.10'];
  }

  let oldestSupportedRelease: semver.SemVer | null = null;
  let newestSupportedRelease = semver.parse(apiCompatibilityData.maximumVersion + '.0')!;

  for (const [releaseVersionString, releaseData] of Object.entries(releases)) {
    const releaseVersion = semver.parse(releaseVersionString + '.0');
    if (!releaseVersion || semver.lt(releaseVersion, FIRST_SUPPORTED_RELEASE)) {
      continue;
    }

    if (semver.gt(releaseVersion, newestSupportedRelease)) {
      const featureFreezeDate = new Date(releaseData.feature_freeze);
      const twoWeeksFromNow = new Date();
      twoWeeksFromNow.setDate(twoWeeksFromNow.getDate() + 14);
      
      if (featureFreezeDate < twoWeeksFromNow) {
        newestSupportedRelease = releaseVersion;
      }
    }

    const endOfLifeDate = new Date(releaseData.end);
    // The GHES version is not actually end of life until the end of the day specified by
    // `end_of_life_date`. Wait an extra week to be safe.
    const oneWeekAfterEol = new Date(endOfLifeDate);
    oneWeekAfterEol.setDate(oneWeekAfterEol.getDate() + 7);
    const isEndOfLife = new Date() > oneWeekAfterEol;
    
    if (!isEndOfLife && (oldestSupportedRelease === null || semver.lt(releaseVersion, oldestSupportedRelease))) {
      oldestSupportedRelease = releaseVersion;
    }
  }

  if (!oldestSupportedRelease) {
    throw new Error('No oldest supported release found');
  }

  const updatedApiCompatibility: ApiCompatibility = {
    minimumVersion: `${oldestSupportedRelease.major}.${oldestSupportedRelease.minor}`,
    maximumVersion: `${newestSupportedRelease.major}.${newestSupportedRelease.minor}`,
  };

  fs.writeFileSync(API_COMPATIBILITY_PATH, JSON.stringify(updatedApiCompatibility, null, 2) + '\n');
}

if (require.main === module) {
  main();
}